compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/ExpectedTypeFinder.java

```diff
@@ -0,0 +1,1138 @@
+            return Optional.empty();
+        }
+
+        if (!(bLangNode instanceof BLangArrowFunction)) {
+            return Optional.empty();
+        }
+
+        BType returnType = ((BLangArrowFunction) nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange()))
+                .funcType.getReturnType();
+        if (!node.rightDoubleArrow().isMissing() &&
+                node.rightDoubleArrow().lineRange().endLine().offset() <= linePosition.offset()) {
+            // Cursor is at the expression node
+            return Optional.of(typesFactory.getTypeDescriptor(returnType));
+        }
+
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(PositionalArgumentNode node) {
+        BLangNode bLangNode = nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange());
+        if (bLangNode == null) {
+            return Optional.empty();
+        }
+
+        if (bLangNode instanceof BLangRecordLiteral) {
+            return getExpectedType(bLangNode);
+        }
+
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(FunctionDefinitionNode node) {
+        Optional<ReturnTypeDescriptorNode> returnTypeDesc = node.functionSignature().returnTypeDesc();
+        if (returnTypeDesc.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Optional<Symbol> functionSymbol = this.semanticModel.symbol(node);
+
+        if (functionSymbol.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return ((FunctionSymbol) functionSymbol.get()).typeDescriptor().returnTypeDescriptor();
+    }
+
+    public Optional<TypeSymbol> transform(MatchClauseNode matchClauseNode) {
+        return matchClauseNode.parent().apply(this);
+    }
+
+    public Optional<TypeSymbol> transform(MatchStatementNode matchStatementNode) {
+        return this.semanticModel.typeOf(matchStatementNode.condition());
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(MappingMatchPatternNode mappingMatchPatternNode) {
+        return mappingMatchPatternNode.parent().apply(this);
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(ExplicitNewExpressionNode node) {
+        BLangNode bLangNode = nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange());
+        if (bLangNode == null) {
+            return Optional.empty();
+        }
+
+        ParenthesizedArgList parenthesizedArgList = node.parenthesizedArgList();
+        // check its is parameter context
+        if (isWithinParenthesis(linePosition, parenthesizedArgList.openParenToken(),
+                parenthesizedArgList.closeParenToken())) {
+            if (bLangNode instanceof BLangTypeInit) {
+                List<BVarSymbol> params = ((BInvokableSymbol) (((BLangInvocation) ((BLangTypeInit) bLangNode).
+                        initInvocation).symbol)).params;
+                if (params.size() == 0) {
+                    throw new IllegalStateException();
```

Comment:
Shall we return Optional.empty() instead of throwing an exception? We have not asked the API users to explicitly handle this exception. The LS functionality will be affected due to such exceptions.

Reply:
Here, LS functionality wont be affected by this exception. This exception is handled internally within the semantic API and if the exception occurs it will return Optional.empty().

Here Optional.empty() can not be used because if there is Optional.empty() we search for the expected type in the node parent. So we can not return Optional.empty() in invalid cases.

Reply:
I see. Let's proceed with the current approach then